### PR TITLE
Unit Test Case Generation

### PR DESCRIPTION
**Pull request by sree24lakshmi**
Revert find_langchain_model changes

### CODE REVIEW

**Comment by rawkintrevo** [comment]
@fayvor possibly of interest to you too

**Review by deanwampler — COMMENTED** [review]
This probably can't be fixed, the indentation of the generated code is terrible. Sometimes there is no indentation when I run it, which is obviously bad. Most of the time, one space is used, but only at the first level. Also, the tests themselves are very underwhelming. At best they show you the basics of how to use an API, but the user will have a lot of work to do afterwards.

**Comment by sree24lakshmi** [comment]
Hi Dean

I agree and have raised the indentation concern with Adam prior to submitting a pull request. As part of the recent code changes that includes invocation of model via langchain replicate directly instead of utils package module, the indentation was retained. The latest code has been uploaded. Our understanding has been to show the generic usage of the model to generate use cases using different test libraries. Please advise if you need us to include any additional use cases/test criteria. Kindly review and share if any further feedback.

**Comment by fayvor** [comment]
@deanwampler @sree24lakshmi I wouldn't put too much effort into adjusting the prompt to get the right output indentation. As long as the code works, you can run it through a formatter (eg. [black](https://github.com/psf/black)). 

In fact, you could do that in the recipe. You can prompt the LLM to output only code, and use this:
```
import black
formatted_code = black.format_str(code_string, mode=black.FileMode())
```

**Comment by fayvor** [comment]
@sree24lakshmi is it worth adding a re-try when the generated python code isn't executable? As folks are learning with AI agents, fault-tolerance has to be built-in to production systems. Retries are one approach to this. Maybe something we can demonstrate here?

**Comment by sree24lakshmi** [comment]
Sure, @fayvor. We could do that but I believe the indentation issue is fixed in the latest code.  Our observation has been that missing indentation scenario was encountered while invoking the model via langchain utils repo package. As per the latest code, with the direct call to langchain replicate model, this no longer seems to be the case as the indented code is returned always.

File: recipes/UnitTest_Code_Gen/unit_test_cases_gen_new.ipynb [hunk 1]
```diff
@@ -0,0 +1,859 @@
...
+   "cell_type": "code",
+   "execution_count": 7,
+   "id": "001fe1b3",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Granite response from Replicate: Here is the unit test code using pytest library package for the given code:\n",
```
[the remark this answers is not in the document]
Comment (deanwampler) [reply recipes/UnitTest_Code_Gen/unit_test_cases_gen_new.ipynb:433]: I'll give you a PR that adds the instructions in the prompt to mock out Kafka.

**Review by deanwampler — CHANGES_REQUESTED** [review]
I'll give you a PR for this branch with my suggested changes. If you are okay with them, more or less, then I'm ready to approve this PR.

File: recipes/UnitTest_Code_Gen/unit_test_cases_gen_new.ipynb [hunk 2]
```diff
@@ -0,0 +1,859 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "79545db2",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#!python -m pip install langchain_community replicate"
```
[the remark this answers is not in the document]
Comment (sree24lakshmi) [reply recipes/UnitTest_Code_Gen/unit_test_cases_gen_new.ipynb]: This is done

File: recipes/UnitTest_Code_Gen/unit_test_cases_gen_new.ipynb [hunk 3]
```diff
@@ -0,0 +1,859 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "54f7fa00",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#!python -m pip install git+https://github.com/ibm-granite-community/utils"
```
[the remark this answers is not in the document]
Comment (sree24lakshmi) [reply recipes/UnitTest_Code_Gen/unit_test_cases_gen_new.ipynb]: Done

File: recipes/UnitTest_Code_Gen/unit_test_cases_gen_new.ipynb [hunk 4]
```diff
@@ -0,0 +1,859 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "id": "b7432e40",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from ibm_granite_community.langchain_utils import set_env_var, get_env_var"
```
[the remark this answers is not in the document]
Comment (sree24lakshmi) [reply recipes/UnitTest_Code_Gen/unit_test_cases_gen_new.ipynb]: This is done as well

File: recipes/UnitTest_Code_Gen/unit_test_cases_gen_new.ipynb [hunk 5]
```diff
@@ -0,0 +1,859 @@
...
+    "\n",
+    "Obtain your REPLICATE_API_TOKEN at replicate.com/account/api-tokens\n",
+    "\n",
+    "There are three ways to provide this value to the cells below. In order of precedence:\n",
+    "\n",
+    "1. As an environment variable\n",
+    "2. As a Google colab secret\n",
+    "3. Supplied by the user using getpass()\n",
+    "\n",
+    "Here: Created a  environment variable `REPLICATE_API_TOKEN='xxxxx'` in `.env` file in the current directory."
```
[the remark this answers is not in the document]
Comment (sree24lakshmi) [reply recipes/UnitTest_Code_Gen/unit_test_cases_gen_new.ipynb]: Implemented the same

File: recipes/UnitTest_Code_Gen/unit_test_cases_gen_new.ipynb (hunk 1 above)
[the remark this answers is not in the document]
Comment (sree24lakshmi) [reply recipes/UnitTest_Code_Gen/unit_test_cases_gen_new.ipynb]: This too has been revised

File: recipes/UnitTest_Code_Gen/unit_test_cases_gen.ipynb [hunk 6]
```diff
@@ -0,0 +1,593 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "54f7fa00",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!pip install git+https://github.com/ibm-granite-community/utils"
```
Comment (fayvor) [inline recipes/UnitTest_Code_Gen/unit_test_cases_gen.ipynb:80]: This and the `pip install langchain_community replicate` above can be pulled in with:
`!pip install git+https://github.com/ibm-granite-community/granite-kitchen`

As shown [here](https://github.com/ibm-granite-community/granite-kitchen/blob/main/recipes/Getting_Started/Getting_Started_with_Replicate.ipynb). You may just want to check that you are getting the same version of replicate.

**Comment by fayvor** [comment]
> multiple retry attempts, rarely yielded successful outcome (indented code).

Makes sense. I'm glad you were able to fix the indentation issue. It's odd to me that the same remote model would return different results, but I guess it's possible that you were using a version of the replicate library that permuted the input or the output.

@sree24lakshmi if you are a member of the IBM Granite org on Replicate, you can look at the outputs server-side by clicking through on your request id [here](https://replicate.com/predictions).

**Comment by deanwampler** [comment]
Thanks for the updates @sree24lakshmi! I'll review shortly.

**Review by deanwampler — APPROVED** [review]
Done

**Comment by deanwampler** [comment]
Done!!

**Comment by sree24lakshmi** [comment]
> > multiple retry attempts, rarely yielded successful outcome (indented code).
> 
> Makes sense. I'm glad you were able to fix the indentation issue. It's odd to me that the same remote model would return different results, but I guess it's possible that you were using a version of the replicate library that permuted the input or the output.
> 
> @sree24lakshmi if you are a member of the IBM Granite org on Replicate, you can look at the outputs server-side by clicking through on your request id [here](https://replicate.com/predictions). That might help ensure no inputs or outputs are getting changed by the client.

Thank you @fayvor

**Comment by sree24lakshmi** [comment]
> Done!!

Thank you @deanwampler for all the support!!

**Pull request merged**